include/gunrock/algorithms/hits.hxx

```diff
@@ -0,0 +1,375 @@
+  using edge_t = typename graph_t::edge_type;
+  using weight_t = typename graph_t::weight_type;
+
+  int iterator = 0;
+  int max_iterations;
+
+  // Sum of squares of authority and hub
+  // qqq find a way to reduce mem access
+  float sum = 0;
+
+  vertex_t vertex_num = 0;
+
+  thrust::device_vector<float> auth_curr;
+  thrust::device_vector<float> hub_curr;
+  thrust::device_vector<float> auth_next;
+  thrust::device_vector<float> hub_next;
+
+  // poniters to the data inside the device_vector
+  float* auth_curr_p = nullptr;
+  float* hub_curr_p = nullptr;
+  float* auth_next_p = nullptr;
+  float* hub_next_p = nullptr;
+
+  // Functor for deviding each element with sum and taking the sqrt
+  template<typename T>
+  struct op{
+  private:
+    float sum;
+  public:
+    op(float sum) : sum(sum){}
+    __device__ __host__
+    T operator()(const T& x) const {
+      return sqrt(x/this->sum);
+      }
+  };
+
+  problem_t(graph_t& G,
+            std::shared_ptr<cuda::multi_context_t> _context,
+            int max_iterations)
+      : gunrock::problem_t<graph_t>(G, _context),
+        max_iterations(max_iterations){
+  vertex_num = G.get_number_of_vertices();
+  auto policy = _context->get_context(0)->execution_policy();
+
+  auth_curr.resize(vertex_num);
+  thrust::fill(policy, auth_curr.begin(), auth_curr.end(), 0);
+  auth_next.resize(vertex_num);
+  thrust::fill(policy, auth_next.begin(), auth_next.end(), 0);
+  hub_curr.resize(vertex_num);
+  thrust::fill(policy, hub_curr.begin(), hub_curr.end(), 0);
+  hub_next.resize(vertex_num);
+  thrust::fill(policy, hub_next.begin(), hub_next.end(), 0);
+  }
+
+  void init() override{
+    auth_curr_p = auth_curr.data().get();
+    hub_curr_p = hub_curr.data().get();
+    auth_next_p = auth_next.data().get();
+    hub_next_p = hub_next.data().get();
+  }
+  void reset() override{}
+
+  bool is_converged(){
+    if(this->max_iterations <= iterator){
+      // qqq notify exceeding
+      return true;
+    }
+    // qqq omit policy
+    else if(thrust::equal(this->auth_curr.begin(),
+            this->auth_curr.end(),
+            this->auth_next.begin())){
+      return true;
+    }
+    else if(thrust::equal(this->hub_curr.begin(),
+            this->hub_curr.end(),
+            this->hub_next.begin())){
+      return true;
+    }
+    else{
+      return false;
+    }
+  }
+
+  void update_iterator(){++this->iterator;}
+
+  // Swap the auth_curr <-> auth_next
+  // and hub_curr <-> hub_next
+  __device__
+  void swap_buffer(){
+    thrust::swap(auth_curr, auth_next);
+    thrust::swap(hub_curr, hub_next);
+  }
+
+  __device__
+  void update_auth(int dest_pos, int source_pos){
+    gunrock::math::atomic::add(&this->auth_next_p[dest_pos],
+                               hub_curr_p[source_pos]);
+  }
+
+  __device__
+  void update_hub(int dest_pos, int source_pos){
+    gunrock::math::atomic::add(&hub_next_p[dest_pos],
+                               auth_curr_p[source_pos]);
+  }
```

Comment:
One line functions? Either use `__force_inline__` or just incorporate the contents of the function as the instruction you want to do. i.e. you do not need these to be separate functions. Just use the atomic add and comment that this updates hub or auth.